From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: =?UTF-8?q?I=C3=B1aki=20Amatria=20Barral?= <inaki.amatria@appentra.com>
Date: Fri, 28 Mar 2025 12:09:30 +0100
Subject: Add semantic accessors to macros

This way we can query the node and implement children-specific logic to
`content` and `alternative`. The nodes below `content` and `alternative`
may be Fortran nodes and not preprocessor directives!! So, for instance,
we want to use `&` as line separator instead of `\`.
---
 grammar.js                   | 10 +++---
 test/corpus/preprocessor.txt | 60 ++++++++++++++++++------------------
 2 files changed, 35 insertions(+), 35 deletions(-)

diff --git a/grammar.js b/grammar.js
index 8b2302e..cfb2254 100644
--- a/grammar.js
+++ b/grammar.js
@@ -2492,7 +2492,7 @@ function preprocIf(suffix, content, precedence = 0) {
       field('condition', $._preproc_expression),
       optional(preprocComment($)),
       '\n',
-      content($),
+      field('content', content($)),
       field('alternative', optional(alternativeBlock($))),
       preprocessor('endif'),
       optional(preprocComment($)),
@@ -2502,7 +2502,7 @@ function preprocIf(suffix, content, precedence = 0) {
       choice(preprocessor('ifdef'), preprocessor('ifndef')),
       field('name', $.identifier),
       optional(preprocComment($)),
-      content($),
+      field('content', content($)),
       field('alternative', optional(alternativeBlock($))),
       preprocessor('endif'),
       optional(preprocComment($)),
@@ -2511,7 +2511,7 @@ function preprocIf(suffix, content, precedence = 0) {
     ['preproc_else' + suffix]: $ => prec(precedence, seq(
       preprocessor('else'),
       optional(preprocComment($)),
-      content($),
+      field('content', content($)),
     )),
 
     ['preproc_elif' + suffix]: $ => prec(precedence, seq(
@@ -2519,7 +2519,7 @@ function preprocIf(suffix, content, precedence = 0) {
       optional(preprocComment($)),
       field('condition', $._preproc_expression),
       '\n',
-      content($),
+      field('content', content($)),
       field('alternative', optional(alternativeBlock($))),
     )),
 
@@ -2527,7 +2527,7 @@ function preprocIf(suffix, content, precedence = 0) {
       choice(preprocessor('elifdef'), preprocessor('elifndef')),
       field('name', $.identifier),
       optional(preprocComment($)),
-      content($),
+      field('content', content($)),
       field('alternative', optional(alternativeBlock($))),
     )),
   };
diff --git a/test/corpus/preprocessor.txt b/test/corpus/preprocessor.txt
index 336057e..b0b35bd 100644
--- a/test/corpus/preprocessor.txt
+++ b/test/corpus/preprocessor.txt
@@ -136,7 +136,7 @@ end subroutine foo4
 (translation_unit
   (preproc_ifdef
     name: (identifier)
-    (subroutine
+    content: (subroutine
       (subroutine_statement
         name: (name)
         (end_of_statement))
@@ -145,34 +145,34 @@ end subroutine foo4
         (end_of_statement))))
   (preproc_ifdef
     name: (identifier)
-    (subroutine
+    content: (subroutine
       (subroutine_statement
         name: (name)
         (end_of_statement))
       (end_subroutine_statement
         (name)
         (end_of_statement)))
-    (preproc_def
+    content: (preproc_def
       name: (identifier)
       value: (preproc_arg))
     alternative: (preproc_elif
       condition: (preproc_defined
         (identifier))
       alternative: (preproc_else
-        (subroutine
+        content: (subroutine
           (subroutine_statement
             name: (name)
             (end_of_statement))
           (end_subroutine_statement
             (name)
             (end_of_statement)))
-        (preproc_def
+        content: (preproc_def
           name: (identifier)
           value: (preproc_arg))))
     (multiline_preproc_comment))
   (preproc_ifdef
     name: (identifier)
-    (subroutine
+    content: (subroutine
       (subroutine_statement
         name: (name)
         (end_of_statement))
@@ -209,7 +209,7 @@ end subroutine foo5
 (translation_unit
   (preproc_ifdef
     name: (identifier)
-    (subroutine
+    content: (subroutine
       (subroutine_statement
         name: (name)
         (end_of_statement))
@@ -218,7 +218,7 @@ end subroutine foo5
         (end_of_statement)))
     alternative: (preproc_elifdef
       name: (identifier)
-      (subroutine
+      content: (subroutine
         (subroutine_statement
           name: (name)
           (end_of_statement))
@@ -227,7 +227,7 @@ end subroutine foo5
           (end_of_statement)))))
   (preproc_ifdef
     name: (identifier)
-    (subroutine
+    content: (subroutine
       (subroutine_statement
         name: (name)
         (end_of_statement))
@@ -236,7 +236,7 @@ end subroutine foo5
         (end_of_statement)))
     alternative: (preproc_elifdef
       name: (identifier)
-      (subroutine
+      content: (subroutine
         (subroutine_statement
           name: (name)
           (end_of_statement))
@@ -244,7 +244,7 @@ end subroutine foo5
           (name)
           (end_of_statement)))
       alternative: (preproc_else
-        (subroutine
+        content: (subroutine
           (subroutine_statement
             name: (name)
             (end_of_statement))
@@ -280,7 +280,7 @@ end subroutine e
 (translation_unit
   (preproc_ifdef
     name: (identifier)
-    (subroutine
+    content: (subroutine
       (subroutine_statement
         name: (name)
         (end_of_statement))
@@ -290,7 +290,7 @@ end subroutine e
     alternative: (preproc_elif
       condition: (preproc_defined
         (identifier))
-      (subroutine
+      content: (subroutine
         (subroutine_statement
           name: (name)
           (end_of_statement))
@@ -300,7 +300,7 @@ end subroutine e
   (preproc_if
     condition: (preproc_defined
       (identifier))
-    (subroutine
+    content: (subroutine
       (subroutine_statement
         name: (name)
         (end_of_statement))
@@ -309,7 +309,7 @@ end subroutine e
         (end_of_statement)))
     alternative: (preproc_elifdef
       name: (identifier)
-      (subroutine
+      content: (subroutine
         (subroutine_statement
           name: (name)
           (end_of_statement))
@@ -317,7 +317,7 @@ end subroutine e
           (name)
           (end_of_statement)))
       alternative: (preproc_else
-        (subroutine
+        content: (subroutine
           (subroutine_statement
             name: (name)
             (end_of_statement))
@@ -348,22 +348,22 @@ General if blocks
         (identifier))
       right: (preproc_defined
         (identifier)))
-    (preproc_def
+    content: (preproc_def
       name: (identifier)
       value: (preproc_arg))
     alternative: (preproc_elif
       condition: (preproc_defined
         (identifier))
-      (preproc_def
+      content: (preproc_def
         name: (identifier))
       alternative: (preproc_elif
         condition: (unary_expression
           argument: (preproc_defined
             (identifier)))
-        (preproc_def
+        content: (preproc_def
           name: (identifier))
         alternative: (preproc_else
-          (preproc_include
+          content: (preproc_include
             path: (system_lib_string)))))))
 
 ================================================================================
@@ -381,7 +381,7 @@ Unary not
     condition: (unary_expression
       argument: (preproc_defined
         (identifier)))
-    (preproc_def
+    content: (preproc_def
       name: (identifier))))
 
 ================================================================================
@@ -500,10 +500,10 @@ end program
       condition: (binary_expression
         left: (identifier)
         right: (number_literal))
-      (use_statement
+      content: (use_statement
         (module_name
           (name)))
-      (end_of_statement))
+      content: (end_of_statement))
     (end_program_statement
       (end_of_statement))))
 
@@ -601,7 +601,7 @@ end module foo
           (end_of_statement)))
       (preproc_ifdef
         name: (identifier)
-        (subroutine
+        content: (subroutine
           (subroutine_statement
             name: (name)
             (end_of_statement))
@@ -661,16 +661,16 @@ end module foo
         (end_of_statement)
         (preproc_ifdef
           name: (identifier)
-          (variable_declaration
+          content: (variable_declaration
             type: (intrinsic_type)
             declarator: (identifier))
-          (end_of_statement)
-          (assignment_statement
+          content: (end_of_statement)
+          content: (assignment_statement
             left: (identifier)
             right: (math_expression
               left: (identifier)
               right: (number_literal)))
-          (end_of_statement))
+          content: (end_of_statement))
         (assignment_statement
           left: (identifier)
           right: (math_expression
@@ -737,11 +737,11 @@ end module foo
         (end_of_statement)
         (preproc_ifdef
           name: (identifier)
-          (assignment_statement
+          content: (assignment_statement
             left: (identifier)
             right: (number_literal
               kind: (identifier)))
-          (end_of_statement))
+          content: (end_of_statement))
         (end_subroutine_statement
           (name)
           (end_of_statement)))
